Consume new `References3.AddFiles` API

tmeschter opened the pull request:
## Bug
Fixes: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/572174
Regression: No
If Regression then when did it last work:  N/A
If Regression then how are we preventing it in future:   N/A

## Fix
Details:

This change adds basic usage of the new `References3.AddFiles` API when
adding references in a packages.config-based project. The current API
used for this purposes, `References.Add`, can trigger multiple
design-time builds per call and may not set all the right metadata on
the resulting `<Reference>` element in the project file. This can
require further calls to update that metadata, which can in turn trigger
further design-time builds.

`AddFiles` on the other hand creates `<Reference>` items with all the
proper metadata--notable, the `<Private>` metadata is set appropriately,
and it specifies an explicit version of the given assembly. And since it
is explicitly about adding files as references and does not need to
support short or full assembly names it can skip several of the
design-time builds run by `Add`. Use of `AddFiles` instead of `Add`
eliminates up to 75% of the design-time builds incurred while adding a
new package.

For the time being, if `AddFiles` fails we will fall back to calling
`Add`. This way adding a new package is not completely broken in case
there are bugs in the `AddFiles` implementation. If the new API proves
reliable, then the fallback can be removed later.

`AddFiles` can also take multiple file paths, and resolve all the
assemblies with a single design-time build. However, using the API in
that manner will require some refactoring of the client code; it is an
option for the future but is not pursued here.

## Testing/Validation
Tests Added: No  
Reason for not adding tests:  Change concerns integration with Visual Studio
Validation done:  Manually verified that with the right changes in place to both the NuGet client and the underlying project system adding new packages works end-to-end.

tmeschter committed:
Consume new `References3.AddFiles` API

This change adds basic usage of the new `References3.AddFiles` API when
adding references in a packages.config-based project. The current API
used for this purposes, `References.Add`, can trigger multiple
design-time builds per call and may not set all the right metadata on
the resulting `<Reference>` element in the project file. This can
require further calls to update that metadata, which can in turn trigger
further design-time builds.

`AddFiles` on the other hand creates `<Reference>` items with all the
proper metadata--notable, the `<Private>` metadata is set appropriately,
and it specifies an explicit version of the given assembly. And since it
is explicitly about adding files as references and does not need to
support short or full assembly names it can skip several of the
design-time builds run by `Add`. Use of `AddFiles` instead of `Add`
eliminates up to 75% of the design-time builds incurred while adding a
new package.

For the time being, if `AddFiles` fails we will fall back to calling
`Add`. This way adding a new package is not completely broken in case
there are bugs in the `AddFiles` implementation. If the new API proves
reliable, then the fallback can be removed later.

`AddFiles` can also take multiple file paths, and resolve all the
assemblies with a single design-time build. However, using the API in
that manner will require some refactoring of the client code; it is an
option for the future but is not pursued here.

tmeschter commented:
Note that this change will not build under the new `VSLangProj157` package has been uploaded to MyGet or NuGet; as such this should not be merged until after that is done.

rohit21agrawal commented:
@tmeschter let me know when you are ready to publish the package, and we will run this through our CI

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:
```diff
@@ -727,6 +727,19 @@ public VSLangProj.References References
+                ThreadHelper.ThrowIfNotOnUIThread();
+
+                dynamic projectObj = VsProjectAdapter.Project.Object;
+                var references = (VSLangProj157.References3)projectObj.References;
```
should we differentiate between NPE vs other exceptions? like if it fails to type cast into `Reference3`

tmeschter replied:
Hmm. What if we reported a telemetry event if the type cast fails? That should only happen if there's a setup authoring issue (e.g. the type library wasn't registered correctly or something like that) or we're encountering a project system that implements `References` but not `References3`. Either way we would want to know.

Does that sound reasonable?

jainaashish replied:
yeah that sounds good.

tmeschter replied:
@jainaashish I've added a very simple telemetry event.

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:
```diff
@@ -794,7 +807,20 @@ public virtual void AddGacReference(string name)
+                    {
+                        // First try the References3.AddFiles API, as that will incur fewer
+                        // design-time builds.
+                        References3.AddFiles(new[] { assemblyFullPath }, out var referencesArray);
```
Shouldn't we now remove setting CopyLocal explicitly? Look at line no 840

tmeschter replied:
If we fall back to `References.Add` we still need to set CopyLocal explicitly.

jainaashish replied:
hmm i see, will there be any perf impact of setting this CopyLocal for `Reference3`?

tmeschter replied:
No, since `TrySetCopyLocal` first checks the `CopyLocal` value, and only sets it if it is not `true`. When using `References3.AddFiles` it will always be `true`.

tmeschter committed:
Telemetry for invalid casts

Report a simple telemetry event if the cast to `References3` fails. This
indicates that we stumbled upon a project system that doesn't implement
this interface yet, or we have a setup authoring where the
VSLangProj157.dll primary interop assembly isn't installed properly.
This way we can find and correct the issues while still handling the
exception and falling back to using the `Reference` interface.

jainaashish reviewed (approved):
Looks good, thanks Tom for the PR.

jainaashish commented:
@tmeschter is the package published?

tmeschter commented:
@jainaashish Not yet. The PR with the project system implementation (including the package .nuspec) has been approved as of today, but I'm waiting until Monday to merge it in on the off chance it breaks something. :-) Once it has gone in I'll pick up the package from the next build.

zhili1208 commented:
@tmeschter does this PR have any VS version requirement? it can't find "VSLangProj157, Version=15.7.0.0" in our EndToEnd test.

rohit21agrawal commented:
@zhili1208 you might want to take a look at how the EndToEnd.zip is created, maybe you can workaround this by including the above DLL directly into the package.

zhili1208 commented:
@rohit21agrawal it's not just EndToEnd test, apex test also failed, looks like API only works in 15.7, just manually tried 15.6, didn't work. we need to update the VS in our CI agent

zhili1208 commented:
@tmeschter is there any way to get 15.6 works with this API?

tmeschter commented:
@zhili1208 No, this depends on a new API that isn't available prior to 15.7.

zhili1208 commented:
merged those two commits in https://github.com/NuGet/NuGet.Client/pull/2133
